Utils/API/server/lib/dkb/api/storages/es.py

```diff
@@ -502,10 +499,15 @@ def get_output_formats(tags):
     :rtype: list
     """
     formats = []
-    for tag in tags:
-        f = OUTPUT_FORMATS.get(tag)
-        if f:
-            formats += f
+    query = dict(TASK_KWARGS)
+    kwargs = {'tags': tags}
+    query['body'] = get_query('output_formats', **kwargs)
+    query['_source'] = ['data_format']
+    query['doc_type'] = 'output_dataset'
+    query['size'] = 500
+    r = client().search(**query)
+    for hit in r['hits']['hits']:
+        formats += hit['_source']['data_format']
```

Comment:
This function takes from ES first 500 output datasets that match the [query](https://github.com/PanDAWMS/dkb/blob/c5326669ce489478a35d8c096152446ba37281a2/Utils/API/server/lib/dkb/api/storages/es/query/output_formats) condition ("dataset has parent with given `ctag`"), and then aggregates received data to get the list. Right?

My questions:
1. Wouldn't it be better to allow ES to make the aggregation?
Why:
    * for now there's a possibility to get incorrect result *(**Q**: when?)*
    * ES can do it faster *(**Q**: why?)*

2. Why only `ctag`?
(In this case) we seem to be interested only in formats for given `ctag-project` combination, and if this list is shorter than that of all formats for given `ctag` -- we're sending extra queries to ES that return no data, yet take some time to send request and get the response. It's not necessary that ES is locally available, so there's gonna be some additional cost for the network operations.

Reply:
> This function takes from ES first 500 output datasets that match the [query](https://github.com/PanDAWMS/dkb/blob/c5326669ce489478a35d8c096152446ba37281a2/Utils/API/server/lib/dkb/api/storages/es/query/output_formats) condition ("dataset has parent with given `ctag`"), and then aggregates received data to get the list. Right?
>

Yes. 

> My questions:
> 
> 1. Wouldn't it be better to allow ES to make the aggregation?

It would, will look into it.

>    Why:
>    
>    * for now there's a possibility to get incorrect result _(**Q**: when?)_

When there is more than 500 documents. As I've said in some other discussion, this may be enough for now, but is not enough in arbitrary case. ES aggregation, on the other hand, goes over all documents.

>    * ES can do it faster _(**Q**: why?)_

Because ES was designed to handle large amounts of data and do various things (including aggregations like this one) with them, while current mechanism is a simple Python code - it is not bad, but it is not specialized as well. The difference would be even more obvious if we were to upgrade the code to deal with the arbitrary case, where all the data had to be retrieved by pages, these pages stored somewhere, then the calculations performed over them...

> 2. Why only `ctag`?
>    (In this case) we seem to be interested only in formats for given `ctag-project` combination, and if this list is shorter than that of all formats for given `ctag` -- we're sending extra queries to ES that return no data, yet take some time to send request and get the response. It's not necessary that ES is locally available, so there's gonna be some additional cost for the network operations.

Also makes sense, will do.

Reply:
It seems that I've overestimated the value generated by the use of aggregation - since we are still retrieving values (output formats) instead of statistics, 'size' limit remains in play. Still, it would take longer to reach 500 different output formats than 500 different datasets with parents with given ctags, and other advantages of the change remain.

Addition of `project` yields no visible result for now (for given `project-ctag` combination).